lsp/nls/src/requests/completion.rs

```diff
@@ -21,42 +29,194 @@ pub fn handle_completion(
+        id: usize,
+    ) -> Option<Vec<Ident>> {
+        linearization.iter().find_map(|item| match item.kind {
+            TermKind::Record(ref fields) if item.id == id => {
+                Some(fields.keys().map(|ident| ident.clone()).collect())
+            }
+            TermKind::Declaration(_, _, ValueState::Known(new_id))
+            | TermKind::Usage(UsageState::Resolved(new_id))
+                if item.id == id =>
+            {
+                find_record_fields(linearization, new_id)
+            }
+            _ => None,
+        })
+    }
 
-    let item = linearization.item_at(&locator);
+    fn find_record_contract(
+        linearization: &Vec<LinearizationItem<Types>>,
+        id: usize,
+    ) -> Option<Contract> {
+        linearization.iter().find_map(|item| match &item.meta {
+            Some(MetaValue { contracts, .. }) if item.id == id => {
+                contracts.iter().find_map(|contract| {
+                    if let Types(AbsType::Record(..)) = contract.types {
+                        Some(contract.clone())
+                    } else {
+                        None
+                    }
+                })
+            }
+            _ if item.id == id => match item.kind {
+                TermKind::Declaration(_, _, ValueState::Known(new_id))
+                | TermKind::Usage(UsageState::Resolved(new_id)) => {
+                    find_record_contract(&linearization, new_id)
+                }
+                _ => None,
+            },
+            _ => None,
+        })
+    }
 
+    /// Extract identifiers from a row type which forms a record.
+    /// Ensure that ty is really a row type.
+    fn extract_ident(ty: &Box<Types>) -> Vec<Ident> {
+        ty.iter_as_rows()
+            .filter_map(|item| match item {
+                RowIteratorItem::Row(ident, _) => Some(ident.clone()),
+                RowIteratorItem::Tail(..) => None,
+            })
+            .collect::<Vec<_>>()
+    }
+
+    let item = linearization.item_at(&locator);
     if item == None {
+        // NOTE: this might result in a case where completion is not working.
+        // This happens when an item cannot be found from the linearization.
         server.reply(Response::new_ok(id, Value::Null));
         return Ok(());
     }
-
     let item = item.unwrap().to_owned();
 
-    let in_scope: Vec<_> = linearization
-        .get_in_scope(&item)
+    Trace::enrich(&id, linearization);
+
+    let trigger = params.context.and_then(|context| context.trigger_character);
+    let in_scope = match trigger {
+        // Record Completion
+        Some(s) if s == "." => {
+            // Get the ID before the .
+            // if the current source is `let var = {a = 10} in var.`
+            // name = "var"
+            let name = text[..start - 1]
+                .chars()
+                .rev()
+                .skip_while(|c| *c == '.') // skip . (if any)
+                .take_while(|c| c.is_ascii_alphabetic())
```

Comment:
This test to find an identifier (if my understanding is correct: this piece of code is looking for the identifier before the `.`) looks suspicious to me.

E.g. `_` is a valid character in some Rust identifiers yet it will not be accepted here. So `foo_bar.` would look for the identifier `bar`, right?

More generally, I would argue that this can be extracted in a pure function: take a string and extract the identifier.

And I also suggest writing some tests to easily check the behavior is as expected. Currently there is no automated way to ensure the behavior is correct right?

Reply:
You're right! Currently `foo_bar` would not get a completion. 

Yes, I think using a pure function would be better.

But for the tests, I don't think there is any part of the LSP that has tests right now; I think it would be better to make another issue for that.

Reply:
We should be matching this for identifiers: 

```rust 
 #[regex("_?[a-zA-Z][_a-zA-Z0-9-']*")]
    Identifier(&'input str)
```

Reply:
> But for the tests, I don't think there is any part of the LSP that has tests right now; I think it would be better to make another issue for that.

What reasons lead you to do it in another PR?

Does it perhaps require a lot of work to create a first test in `./lsp/nls`? Besides the usual:
* Add test lib in `Cargo.toml`
* Add `cargo test` in CI
 
If not, I'd say it's ok to add the first test as part of this PR, because adding feature X and adding tests about feature X belong in the same PR.

If it is a lot of work, I suggest we shortly press pause on this PR, open a new PR to add test infrastructure for the LSP, then rebase this branch and add tests